File: ProjectionService/service/authService.go
package service

import (
	"encoding/json"
	"errors"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"

	apicontract "github.com/draganaF/gled.io/ProjectionService/apiContract"
)

type AuthService struct {
	client *http.Client
}

func NewAuthService() *AuthService {
	return &AuthService{
		client: &http.Client{},
	}
}

func (authService *AuthService) Authorize(token string, roles *[]int) (*apicontract.AuthorizedUser, error) {
	url := os.Getenv("USER_SERVICE_URL") + "/api/authorize"

	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", token)
	req.Header.Add("Accept", "application/json")

	res, err := authService.client.Do(req)
	if err != nil {
		return nil, errors.New("you are not authorized")
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		return nil, errors.New("could not authorize")
	}

	bodyBytes, err := io.ReadAll(res.Body)
	if err != nil {
		log.Fatal(err)
	}
	user := &apicontract.AuthorizedUser{}
	json.Unmarshal(bodyBytes, user)

	value, _ := strconv.Atoi(user.Role)
	if authService.checkRoles(value, roles) {
		return user, nil
	}

	return nil, errors.New("you don't have permission")
}

func (authService *AuthService) checkRoles(role int, roles *[]int) bool {

	for _, r := range *roles {
		if r == role {
			return true
		}
	}

	return false
}

File: RecensionService/src/service/auth_service.rs
use std::string;

use reqwest::StatusCode;

use crate::model::auth::AuthenticatedUser;



pub struct AuthService {
  client: reqwest::blocking::Client
}

impl AuthService {
    
  pub fn new() -> AuthService {
    AuthService {
      client: reqwest::blocking::Client::new()
    }
  }

  pub fn authorize(&self, token: &String, roles: &Vec<String>) -> Result<AuthenticatedUser, String> {

    let response = self.client.get("http://localhost:8083/api/authorize")
      .header("Authorization", token)
      .send()
      .unwrap();
      
    match response.status() {
      StatusCode::OK => {},
      _ => return Err("You are not authorized".to_string())
    };

    let authenticated_user = response.json::<AuthenticatedUser>().unwrap();
    
    if roles.contains(&authenticated_user.Role) {
      Ok(authenticated_user)
    } else {
      Err("You are not authorized.".to_string())
    }
  }

}